### PR TITLE
Restore & add support for legacy syntax

### PR DESCRIPTION
**Pull request by xeho91**
## Dependencies
Requires #181 to be merged first.

### Support for legacy components for backward compatibility:

1. [`<Meta>`](https://github.com/storybookjs/addon-svelte-csf/blob/main/src/components/Meta.svelte)
    - [x] codemod - `<Meta>` to `defineMeta` & insertion into `<script context="module">`
    - [x] tests

1. [`<Story>`](https://github.com/storybookjs/addon-svelte-csf/blob/main/src/components/Story.svelte)
   - [x] tests
   - [x] codemods:
      - [x] transform `let:args` & `let:context` directives to inner `children` snippet block
      - [x] transform `autodocs` prop to `tags={["autodocs"]}`
      - [x] transform `source` by removing when is a shorthand or to `parameters.docs.source.code`
      - [x] transform `template` prop to `children={<template identifier>}`
        
1. [`<Template>`](https://github.com/storybookjs/addon-svelte-csf/blob/main/src/components/Template.svelte)
   - [x] codemods - replace component with snippet block
   - [x] tests

### Add addon options - `StorybookAddonSvelteCsfOptions`

- [x] update addon's Vite plugin - add `pre-transform` hook
- [x] add option `supportLegacy`

### Deprecation warnings, errors, messages

- [x] document new errors 
- [x] Ensure deprecation warnings are visible enough while writing stories code:
      - writing the code _(with TypeScript support and marking `<Template>` as deprecated)_
- [ ] log deprecated warnings at runtime

### Update stories tests

- [x] Using `LegacyTemplate` component
- [x] Using `LegacyStory` component _(and its legacy props)_
- [x] Using `Meta` component

### Update indexer

- [x] Support indexing legacy `export const meta` syntax
- [x] Support indexing legacy `Meta` component

### TODO

- [ ] https://github.com/storybookjs/addon-svelte-csf/pull/186#discussion_r1685643713
- [x] https://github.com/storybookjs/addon-svelte-csf/pull/186#discussion_r1685644016
- [x] https://github.com/storybookjs/addon-svelte-csf/pull/186#discussion_r1683370952
- [x] https://github.com/storybookjs/addon-svelte-csf/pull/186#discussion_r1685313983

---

<details>
  <summary>📦 Published PR as canary version: <code>4.2.0--canary.186.ff8a94d.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/addon-svelte-csf@4.2.0--canary.186.ff8a94d.0
  # or 
  yarn add @storybook/addon-svelte-csf@4.2.0--canary.186.ff8a94d.0
  ```
</details>

### CODE REVIEW

**Comment by xeho91** [comment]
@JReinhold 

I got tests working. I'm encountering the issue with Chromatic CI which I can't figure out. I don't believe it's related to my changes.

What's left is updating documentation. Most importantly change wording where we declared 'removed' legacy API to 'deprecated'. And also, highlighting usage of the option `legacyTemplate`.

And maybe improve test _(stories)_ for legacy API, because I got a feeling you don't want to provide examples for legacy API, right?

File: src/runtime/LegacyMeta.svelte [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<script lang="ts">
```
Comment (JReinhold) [inline src/runtime/LegacyMeta.svelte:1]: All these legacy components could just throw an error in their instance script right?
If they actually exist at runtime, it must be because the user hans't enabled the legacy flag right?

Perhaps a nice documented error.

File: src/utils/error/codemod/index.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    filename: StorybookSvelteCSFError['filename'];
+    attribute: InvalidTemplateAttribute['attribute'];
+  }) {
+    super({ filename });
+    this.attribute = attribute;
+  }
+
+  template(): string {
+    return dedent`
+      'Story' component prop 'template' value must be a text (string) with a reference to existing '<Template>' component id.
```
Comment (JReinhold) [inline src/utils/error/codemod/index.ts:25]:
```suggestion
      'Story' component prop 'template' value must be a string with a reference to existing '<Template>' component id.
```

File: tests/stories/LegacyMeta.stories.svelte [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<script context="module">
+  import { Meta, Story } from '@storybook/addon-svelte-csf';
+
+  import LegacyMeta from './LegacyMeta.svelte';
+</script>
+
```
Comment (JReinhold) [inline tests/stories/LegacyMeta.stories.svelte:6]: The legacy API has never worked without a `<Template>` call, so we should add one here, even though it technically works without it in v5.

File: tests/stories/LegacyStory.stories.svelte [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+   * * _Italic_,
+   * * `Code`.
+   */
+  export const meta = {
+    title: 'LegacyStory',
+    component: LegacyStory,
+    tags: ['autodocs'],
+  } satisfies Meta<LegacyStory>;
+</script>
+
```
Comment (JReinhold) [inline tests/stories/LegacyStory.stories.svelte:22]: Same as above, needs a `<Template>` here.

File: tests/stories/LegacyStory.stories.svelte [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  <LegacyStory {...args} />
+</Story>
+
+<Story name="With autodocs" autodocs />
+
+<Story name="With source as shorthand" source />
+
+<Story name="With source as text" source="<LegacyStory>{'Hi'}</LegacyStory>">
+  <LegacyStory>{'Hi'}</LegacyStory>
+</Story>
```
Comment (JReinhold) [inline tests/stories/LegacyStory.stories.svelte:42]: Looks like the source snippet is wrong here. See:

image

File: tests/stories/LegacyTemplate.stories.svelte [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+</Template>
+
+<Template id="rounded" let:args>
+  <p>Using rounded template</p>
+  <LegacyTemplate {...args} onclick={handleClick} rounded />
+</Template>
+
+<Story name="Default" />
+
+<Story name="Rounded" template="rounded" />
```
Comment (JReinhold) [inline tests/stories/LegacyTemplate.stories.svelte:43]: I think _this_ is causing the Chromatic failure. I expect you'll see the same if you build the Storybook locally and serve it, maybe even in dev mode too.

This is the raw story as opened in an iframe: https://6042cb92fd1bb200234586ee-poqxheiipz.chromatic.com/iframe.html?id=tests-legacytemplate--default&viewMode=story

First error:

image

Leading to this:

image

I think this is the mismatch in snippet reference that I've commented on earlier. The snippet will be named `rounded()`, but after the codemod the Story will reference it as `children={Rounded}` - and `Rounded` is coincidentally the export name of the story, so it's in fact not referencing the snippet but the story defined as `const Rounded = stories[...]` after post-transformation.

**Review by JReinhold — CHANGES_REQUESTED** [review]
Great job! 👏 Added a lot of comments. The biggest thing here is probably that it doesn't handle stories written exclusively with the instance-script, which we need to support and test.

> And maybe improve test (stories) for legacy API, because I got a feeling you don't want to provide examples for legacy API, right?

Correct. I think the test stories you have for the legacy components are good - and they already caught a bug ;) - no need to add anything to the examples dir.
You could add a more "advanced" test story like the Button one we have in v4 currently: https://github.com/storybookjs/addon-svelte-csf/blob/main/stories/button.stories.svelte

File: tests/stories/LegacyTemplate.stories.svelte (hunk 6 above)
Comment (xeho91) [reply tests/stories/LegacyTemplate.stories.svelte:43]: Haha, I shot myself in the foot with this one. I was troubled with how the hell, then I remember:

https://github.com/storybookjs/addon-svelte-csf/blob/89bd7f1375c391f76aec40390e66e3c0b73fbae2/src/compiler/pre-transform/codemods/legacy-story.ts#L231-L238


Thanks for this amazing work, and I'm going to make a note to myself. Ensure that runtime works by actually manually verifying the runtime 😅

File: src/runtime/LegacyMeta.svelte (hunk 1 above)
Comment (xeho91) [reply src/runtime/LegacyMeta.svelte:1]: Very good point, you're right.

I think adding a checker inside the `index.ts` while importing the legacy components is hard _(no idea if there's a simple way to access addon options)_.
So I can do it while traversing on the Svelte AST inside the indexer parser and post-transform maybe _(because pre-transform will be skipped)_.

File: src/compiler/pre-transform/index.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+      next(state);
+    },
+
+    Root(node, context) {
+      const { fragment, module, ...rest } = node;
+      const { state, visit } = context;
+
+      if (!module) {
+        throw new MissingModuleTagError(filename);
```
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/index.ts:42]: Oh, that complicates the work a bit, I think.

Two solutions:
1. Codemod add the `context="module"` to the `<script>` tag.
   Might be very dangerous operation, because there might be some logic created by the user unrelated to the Storybook usage. For example `onclick` snippets for button tests, etc.
2. Codemod will extract the most important pieces:
    1. import declarations for this addon - `"@storybook/addon-svelte-csf"`
    1. import declaration of the component _(optional)_
    1. `export const meta`
  ... and create a separate `<script context="module">` above the instance tag, with the extracted AST nodes from the instance tag.

File: src/compiler/pre-transform/index.ts [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+      if (node.context === 'module') {
+        next(state);
+      }
+    },
+
+    Program(_node, context) {
+      const { next, state } = context;
+
+      next(state);
+    },
```
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/index.ts:67]: This instructs the AST walker - `zimmerframe` to go "deeper" in the tree. I'm not sure if this can be omitted, but I can test.

`Root` -> `module` _(`Script`)_ -> `Program` -> `nodes` _(declarations)_.

File: src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import pkg from '@storybook/addon-svelte-csf/package.json' with { type: 'json' };
+import type { Comment, Component } from 'svelte/compiler';
+import { print } from 'svelte-ast-print';
+import { describe, it } from 'vitest';
+
+import { transformComponentMetaToDefineMeta } from './component-meta-to-define-meta';
+
+import { parseAndExtractSvelteNode } from '#tests/extractor';
+
+describe(transformComponentMetaToDefineMeta.name, () => {
```
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts:9]: 1. Done in fdebf5ca6486d5e08faef98fede1176156a1be2e
2. This one is tough.
   I'm not sure what should be expected outcome. We move this variable declaration to module tag?
   Currently it is transforming to:

   ```ts
   const { Story } = defineMeta({
      	component: WithParameters,
      	parameters: { ...parameters, baz: 'yes' }
   });
   ```
   which of course would produce an incorrect behaviour at runtime.
   
   This case might need more work. Because apart of `parameters` prop, there could be other properties _(from meta)_ being referenced in variables, e.g. `tags`, `play`, etc.

File: src/compiler/pre-transform/codemods/import-declaration.test.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import pkg from '@storybook/addon-svelte-csf/package.json' with { type: 'json' };
+import type { ImportDeclaration } from 'estree';
+import { print } from 'svelte-ast-print';
+import { describe, it } from 'vitest';
+
+import { transformImportDeclaration } from './import-declaration';
+
+import { parseAndExtractSvelteNode } from '#tests/extractor';
+
+describe(transformImportDeclaration.name, () => {
```
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/codemods/import-declaration.test.ts:10]: Correct, instance-level imports are not considered right now. I think I can extract the import declaration from instance, and put in the module tag.

File: tests/stories/LegacyStory.stories.svelte (hunk 5 above)
Comment (xeho91) [reply tests/stories/LegacyStory.stories.svelte]: Fixed.

It happened because, I overlooked the that Svelte's compiler tries to compute what is inside the curly brackets in text expression value of attribute - in this case `{'Hi'}`

File: src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (JReinhold) [reply src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts:9]: I'm thinking either:

1. we attempt to move it to the module script. This can be tricky, because moving it could require moving _other_ variables as well that it references. This is very edgy though, so maybe not something we need to support.
2. or we detect the reference and log a warning to the user that the legacy API-usage couldn't be supported. Ignore the file completely, tell them to migrate the file to the new API.

File: src/compiler/pre-transform/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JReinhold) [reply src/compiler/pre-transform/index.ts:42]: I think option 2 sounds best, if you think it's doable.

This is crucial though, as I think most usage of the v4 API is with the instance script today.

File: src/runtime/LegacyMeta.svelte (hunk 1 above)
Comment (JReinhold) [reply src/runtime/LegacyMeta.svelte:1]: > no idea if there's a simple way to access addon options

It's doable, but not simple so let's try to avoid it.

> I can do it while traversing on the Svelte AST inside the indexer parser and post-transform maybe (because pre-transform will be skipped without having this option enabled).

Just making the components throw an error should be the simplest solution here right? Like

```svelte
<script>
  throw new Error("you have not enabled the legacy option...");
</script>
```

Comment (xeho91) [reply src/runtime/LegacyMeta.svelte:1]: That's much simpler, thanks for the hint!

Do we need to have test for it? No idea how we could do it in this repo _(with one Storybook config)_ and not breaking the CI(?)

Comment (JReinhold) [reply src/runtime/LegacyMeta.svelte:1]: Not easily no, I'm fine with this just having been tested manually.

File: src/compiler/pre-transform/index.test.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      </script>
+
+      <!-- This is a description for **default** Button. -->
+      <Story name="Default" />"
+    `);
+  });
+
+  it("replaces 'Meta' with inserting 'defineMeta' to module tag", async ({ expect }) => {
+    const code = dedent(`
+      <script context="module">
```
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts:44]: Let me know if those tests got it covered:
- https://github.com/storybookjs/addon-svelte-csf/blob/7aae06f7838357739a73da1ba42074d22c275071/src/compiler/pre-transform/index.test.ts#L144-L158
- https://github.com/storybookjs/addon-svelte-csf/blob/7aae06f7838357739a73da1ba42074d22c275071/src/compiler/pre-transform/index.test.ts#L160-L186
- https://github.com/storybookjs/addon-svelte-csf/blob/7aae06f7838357739a73da1ba42074d22c275071/src/compiler/pre-transform/index.test.ts#L188-L220

If yes, mark this as resolved 🙏

File: src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (xeho91) [reply src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts:9]: 1. This is too much complex work. I can see it when I just finished initial support of moving from instance tag to module:
   - addon import declaration _(and transforming it)_
   - optional stories component import
   - export const meta _(and transforming it)_
   
   I'd have to walk on `ObjectExpression` on `export const meta` or `<Meta>` attributes _(props)_, and finding identifiers to each meta entry at the root level _(depth = 0)_. Now, imagine some very edge cases with depth level > 1 entries _(e.g. for `parameters`)_ located in the instance tag. 😅
   
2. I need to think on how to display this warning - as in how to detect references.

File: src/runtime/LegacyMeta.svelte (hunk 1 above)
Comment (xeho91) [reply src/runtime/LegacyMeta.svelte:1]: In commit: 34d0eacc633683e850a663b1e65b7734f01658cd

I added this error, but after testing manually I don't think this was the right direction 😭 

The first error I got is:
```
storybook v8.2.5

info => Cleaning outputDir: storybook-static
info => Loading presets
info => Building manager..
info => Manager built (116 ms)
info => Building preview..
WARN 🚨 Unable to index ./tests/stories/LegacyWithInstance.stories.svelte:
WARN   SB_SVELTE_CSF_PARSER_EXTRACT_SVELTE_0001 (MissingModuleTagError): The file 'file:///Users/xeho91/Nextcloud/Projects/oss/addon-svelte-csf/tests/stories/LegacyWithInstance.stories.svelte'
WARN does not have a module context (<script context="module"> ... </script>).
WARN
WARN defineMeta(...) should be called inside a module script tag, like so:
WARN
WARN <script context="module">
WARN import { defineMeta } from "@storybook/addon-svelte-csf";
WARN
WARN const { Story } = defineMeta({});
WARN </script>
WARN
WARN More info: https://github.com/storybookjs/addon-svelte-csf/blob/v4.1.2/ERRORS.md#SB_SVELTE_CSF_PARSER_EXTRACT_SVELTE_0001
```

Which makes sense, because indexer parser is being run first. So I guess we have to move error to this parser, which would benefit us by giving us `filename` path.

Comment (xeho91) [reply src/runtime/LegacyMeta.svelte:1]: Your idea of using try catch and recognise known errors from indexer parser related to the legacy API _(https://github.com/storybookjs/addon-svelte-csf/commit/95a4becc4191af91533823a45a7047616d0c4b29)_ does the job - tested manually, thanks! 🙏

**Comment by xeho91** [comment]
@JReinhold PR is ready for review.
Phew, there were some issues so hard to spot, because of missing type support from `svelte/compiler` on their AST nodes _([reference]( https://github.com/sveltejs/svelte/pull/12967))_

Two requests to put attention to:
1. If you could find any spots where we can emit warnings about using legacy syntax, it'll be awesome!
2. https://github.com/storybookjs/addon-svelte-csf/pull/186#discussion_r1685643713 This one is too hard for me to figure out at this moment. As discussed in private, is better left to user to fix very advanced cases on their own.

File: src/compiler/pre-transform/codemods/legacy-story.test.ts [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    expect(print(transformLegacyStory({ node }))).toMatchInlineSnapshot(
+      `
+			"<Story name="Default" parameters={{
+				docs: { source: { code: "'<Button primary />'" } }
+			}} />"
+		`
+    );
+  });
+
+  it("'source' prop when is a text expression gets moved to 'parameters' prop", async ({
```
Comment (JReinhold) [inline src/compiler/pre-transform/codemods/legacy-story.test.ts:76]: duplicate test name

```suggestion
  it("'source' prop when is a text expression and Story has children, gets moved to 'parameters' prop", async ({
```

File: src/compiler/pre-transform/codemods/legacy-story.test.ts [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+    expect(print(transformLegacyStory({ node }))).toMatchInlineSnapshot(`
+			"<Story name="Default">
+				{#snippet children(_args, context)}
+					<p>{context.id}</p>
+				{/snippet}
+			</Story>"
+		`);
+  });
+});
```
Comment (JReinhold) [inline src/compiler/pre-transform/codemods/legacy-story.test.ts:195]: can we add a test more that uses both `args` and `context` directives?

File: src/compiler/pre-transform/codemods/legacy-story.test.ts [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+      <Story name="Default" autodocs />
+    `;
+    const node = await parseAndExtractSvelteNode<Component>(code, 'Component');
+
+    expect(print(transformLegacyStory({ node }))).toMatchInlineSnapshot(
+      `"<Story name="Default" tags={["autodocs"]} />"`
+    );
+  });
+
+  it("moving 'autodocs' prop doesn't break with existing 'tags' prop", async ({ expect }) => {
```
Comment (JReinhold) [inline src/compiler/pre-transform/codemods/legacy-story.test.ts:25]: can we add two more tests to support what we discussed in https://github.com/storybookjs/addon-svelte-csf/pull/207#discussion_r1739989511:

1. `tags="singular"` is converted to `tags=["singular"]`
2. `autodocs tags="singular"` is converted to `tags=["autodocs", "singular"]`

File: src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    const component = await parseAndExtractSvelteNode<Component>(code, 'Component');
+
+    expect(print(transformComponentMetaToDefineMeta({ component }))).toMatchInlineSnapshot(`
+      "const { Story } = defineMeta({
+      	component: WithParameters,
+      	parameters: { ...parameters, baz: 'yes' }
+      });"
+    `);
+  });
+});
```
Comment (JReinhold) [inline src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts:111]: can we add two more tests as per https://github.com/storybookjs/addon-svelte-csf/pull/207#discussion_r1739989511:

1. `<Meta tags="singular" />` -> `tags: ["singular"]`
1. `<Meta tags=["first", "second"] />` -> `tags: ["first", "second"]`

File: src/compiler/pre-transform/index.test.ts [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+				import { defineMeta } from "@storybook/addon-svelte-csf";
+
+				/** This is a description for the **Button** component stories. */
+				const { Story } = defineMeta({ title: "Atoms/Button", component: Button });
+			</script>
+
+			{#snippet children(args)}
+				<Button {...args} />
+			{/snippet}
+			<Story name="Default" />"
```
Comment (JReinhold) [inline src/compiler/pre-transform/index.test.ts:95]: this looks incorrect? `children` isn't referenced

File: src/compiler/pre-transform/index.ts [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+        instance: transformedInstance,
+        module: transformedModule,
+      };
+    },
+
+    // NOTE: It walks either on module or instance tag
+    Script(node, context) {
+      const { next, state } = context;
+
+      state.currentScript = state.currentScript = node.context === 'module' ? 'module' : 'instance';
```
Comment (JReinhold) [inline src/compiler/pre-transform/index.ts:70]:
```suggestion
      state.currentScript = node.context === 'module' ? 'module' : 'instance';
```

File: src/compiler/pre-transform/index.ts [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+      const { declaration } = node;
+      const { state } = context;
+
+      if (
+        declaration &&
+        declaration.type === 'VariableDeclaration' &&
+        declaration.declarations[0].type === 'VariableDeclarator' &&
+        declaration.declarations[0].id.type === 'Identifier' &&
+        declaration.declarations[0].id.name === 'meta'
+      ) {
```
Comment (JReinhold) [inline src/compiler/pre-transform/index.ts:113]: nit: early return

```suggestion
      if (!(
        declaration &&
        declaration.type === 'VariableDeclaration' &&
        declaration.declarations[0].type === 'VariableDeclarator' &&
        declaration.declarations[0].id.type === 'Identifier' &&
        declaration.declarations[0].id.name === 'meta'
      )) {
        return;
      }
```

File: src/index.test.ts [hunk 19]
```diff
@@ -1,5 +1,5 @@
-import type { PlayFunctionContext } from '@storybook/types';
-import type { Component, ComponentProps, Snippet } from 'svelte';
+import type { PlayFunctionContext } from 'storybook/internal/types';
```
Comment (JReinhold) [inline src/index.test.ts:1]: Let's revert this so we can keep supporting 8.0, if possible.

**Review by JReinhold — CHANGES_REQUESTED** [review]
Very impressive work!

File: ERRORS.md [hunk 20]
```diff
@@ -325,3 +330,37 @@ eg. `Some story name!!` will be converted to `SomeStoryName`.
 You can fix this collision by providing a unique `exportName` prop with`<Story exportName="SomeUniqueExportName" ... />`.
 
 See more in [the `exportName` API docs](./README.md#custom-export-name).
+
+## `CODEMOD`
+
+### `SB_SVELTE_CSF_CODEMOD_0001`
```
[the remark this answers is not in the document]
Comment (xeho91) [reply ERRORS.md:336]: Agreed, done

File: src/compiler/pre-transform/codemods/legacy-story.test.ts (hunk 13 above)
Comment (xeho91) [reply src/compiler/pre-transform/codemods/legacy-story.test.ts]: 2468f04cc9f9

File: src/compiler/pre-transform/codemods/legacy-story.test.ts (hunk 14 above)
Comment (xeho91) [reply src/compiler/pre-transform/codemods/legacy-story.test.ts:25]: I will, I need to create a codemod for this one.
I'll do it tomorrow, because it is a wrap for me today.

File: src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts (hunk 15 above)
Comment (xeho91) [reply src/compiler/pre-transform/codemods/component-meta-to-define-meta.test.ts:160]: 09e65099b1fc810b3bc8bb5981133cfa41ac7fb6 and 3c361d487720e0bc213cf046bd218015cdbc7576 should cover it.

File: src/compiler/pre-transform/codemods/legacy-story.test.ts (hunk 14 above)
Comment (xeho91) [reply src/compiler/pre-transform/codemods/legacy-story.test.ts:25]: 09e65099b1fc810b3bc8bb5981133cfa41ac7fb6 and 3c361d487720e0bc213cf046bd218015cdbc7576 should cover it.

File: src/compiler/pre-transform/index.test.ts (hunk 16 above)
Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts:95]: You're right, is totally incorrect. And I'm puzzled on what to do with this case.

Reminder:
1. Component `<Template>` attribute `id` is optional
2. `<Story />` component(s) by default were using the `default` template _(if existed)_, but the reference isn't visible on the stories file _(is a runtime mechanism)_

In this case, is easy to assume that the best solution to this problem is to achieve the following result:

```svelte
<script context="module">
import { defineMeta, setTemplate } from "@storybook/addon-svelte-csf";
/** This is a description for the **Button** component stories. */
const { Story } = defineMeta({ title: "Atoms/Button", component: Button });
</script>

<script>
setTemplate(template);
</script>

{#snippet template(args)}
   <Button {...args} />
{/snippet}

<Story name="Default" />

<Story name="Other" />
```

To create a codemod for this one complicates the logic a bit because:
1. I need to create a `<script>` instance
2. I need to add `setTemplate` to the import declaration for this addon inside the `<script module>` tag.

Do you see a simpler approach to this case?

Comment (JReinhold) [reply src/compiler/pre-transform/index.test.ts:95]: 1. When you find a `<Template/>` without an explicit id, convert it to a snippet with a namespaced name so it won't collide with others, like `{#snippet sb_default_template_1()}`. Also give it a number, in the case where multiple id-less Templates are found, they can have number 1, 2 3... I don't know why anyone would have multiple id-less Templates, but just to be on the safe side?
2. When you find a `<Story/>` without an explicit template reference, make it reference the first default, eg. `children={sb_default_template_1}`

Then you don't need the `setTemplate` complexity at all?

Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts]: I did give it a try. cfc7beb82f63157e791470f73cfb1a5efd4e3c8b

Had to refactor a little bit by passing down the state from `zimmerframe` walk down to transformation function.

I only found one gotcha in `2.` point. Assuming someone did create a bizarre stories _(invalid, but AFAIK has a fallback mechanism behaviour)_, like this:

```svelte
      <Template let:context>
        <p>{context.args}</p>
      </Template>

      <Story name="Default" />

      <Template let:args>
        <Button {...args} />
      </Template>

      <Story name="NextOne" />
```

... which children should the second story `NextOne` refer to? I assumed it should match the latest one. And the tests shows current ouput:

```svelte
      {#snippet sb_default_template_1(_args, context)}
      	<p>{context.args}</p>
      {/snippet}

      <Story name="Default" children={sb_default_template_1} />

      {#snippet sb_default_template_2(args)}
      	<Button {...args} />
      {/snippet}

      <Story name="NextOne" children={sb_default_template_2} />
```

Hope that's ok. Regardless, this case is very edgy, hope it never happens. It was a little eye-opening, because I see why sometimes there's a need to be more conservative with API design. The previous API design is... too *flexible* 😅

Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts]: Argh, there's an edge case that comes with the currently implemented solution. I had to change the components order: c2b6b827913a

Explanation: CI was failing, because `<Story name="default" />` were transformed into `<Story name="default" children={sb_default_template_2} />`. And the reference `sb_default_template_2` isn't defined because the second `<Template>` has an `id` which is used as an identifier for its own transformation to snippet.

I could probably filter out '*id-less*' templates, so this bizarre edge case doesn't happen. But I'll put it on hold until I get to know your opinion on this solution.

Comment (JReinhold) [reply src/compiler/pre-transform/index.test.ts]: Alright, I tested out what the behavior in v4 was - because we should just mimic that. With the follow stories:

```svelte
<Template let:args>
	<Button {...args}>
		First template
	</Button>
</Template>

<Story name="First Story" args={{ rounded: true }} />

<Template let:args>
	<Button {...args}>
		Second template
	</Button>
</Template>

<Story name="Second Story" source args={{ rounded: false }} />
```

You get these two, super weird stories:

image

image

So it seems that if you have multiple templates without `id`, they just combine into a single render?! I cannot fathom why users would ever want this explicitly, and they'd also notice this right away because the whole story renders double. So there's 0.0001 % chance this will ever happen. The same thing happens btw if multiple Templates have the same id.

We don't need to support multiple templates without id/with the same id. Ideally we'd throw a nice error if we detect this, but if that's too hard, then don't even waste time on it, just let it fail however it does naturally.

Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts]: Huh. It might be some issue with logic on the runtime scope of this addon.

Regardless, I see a way to throw an error if there's more than two *'id-less'* `<Template>` components and it certainly feels like an easier solution for this edge scenario.

Comment (xeho91) [reply src/compiler/pre-transform/index.test.ts]: b1bf0c3b665e
